items: add Item_TestFrameRange function

walkawayy opened the pull request:
#### Checklist

- [X] I have read the [coding conventions](https://github.com/LostArtefacts/TR1X/blob/master/CONTRIBUTING.md#coding-conventions)
- [X] I have added a changelog entry about what my pull request accomplishes, or it is an internal change

#### Description
Adds a function to test if an item's frame number is in an inclusive range. This PR also renames `Item_TestFrame` to `Item_TestFrameEqual` in a separate commit in order to better differentiate the two functions.

As part of this refactor, I noticed there are some parts of the code like in `Item_GetAnimChange` where the `Item_SwitchToAnim` function is not used:
```
                    item->anim_number = range->link_anim_num;
                    item->frame_number = range->link_frame_num;
```
This is probably because `Item_SwitchToAnim` uses `frame_base`, but I wanted to note that there are still some areas of the code not using these helper functions which could affect variable FPS down the road if it's ever implemented.

walkawayy committed:
items: rename Item_TestFrame to Item_TestFrameEqual

walkawayy committed:
items: add Item_TestFrameRange function

Resolves #1011.

Fixes Lara not stepping back down one click with her right foot.
Resolves #1014.

rr- commented on src/game/items.c:
```diff
@@ -764,7 +764,12 @@ void Item_TakeDamage(ITEM_INFO *item, int16_t damage, bool hit_status)
+
+bool Item_TestFrameRange(ITEM_INFO *item, int16_t start, int16_t end)
+{
+    return item->frame_number >= start && item->frame_number <= end;
```
I think both functions should take frames relative to the animation base so that they have the same contract.

rr- reviewed (commented):
Can we list the places where we still mess directly with animation numbers rather than using the getter/setter proxy functions?

walkawayy commented:
Added the fix for #1014 but that will need to be tested.

Added constants for the hard coded frame checks. Let me know about the names. It was hard to name the frames in `lara_col.c` especially. Also a funny thing was `torso->flags` was basically being used as `base_frame` so it is no longer needed since `Item_TestFrameRange` uses the `base_frame` already.

This PR covers all hard coded frame range checks which now use `Item_TestFrameRange`.

walkawayy commented:
This post will note remaining hard coded frame equal checks that do not use `Item_TestFrameEqual`:
`Lara_Animate` and `Item_Animate`:
```
            case AC_EFFECT:
                if (item->frame_number == command[0]) {
                    g_EffectRoutines[command[1]](item);
                }
                command += 2;
                break;
            }
```

walkawayy commented:
This post will cover remaining hard coded animation switches that do not use `Item_SwitchToAnim`:

- [ ] `Item_Animate`:

```
        item->anim_number = anim->jump_anim_num;
        item->frame_number = anim->jump_frame_num;
```

- [ ] `Item_GetAnimChange`:
```
                    item->anim_number = range->link_anim_num;
                    item->frame_number = range->link_frame_num;
```

- [x] `Object_DrawPickupItem`:
```
    item->anim_number = g_Objects[item_num_option].anim_index;
    item->frame_number = g_Anims[item->anim_number].frame_base;
```

- [x] `Game_Cutscene_InitialiseHair`:
```
    g_LaraItem->anim_number = g_Objects[lara_type].anim_index;
    ANIM_STRUCT *cut_anim = &g_Anims[g_LaraItem->anim_number];
    g_LaraItem->frame_number = cut_anim->frame_base;
```

- [ ] `Lara_Animate`:
```
        item->anim_number = anim->jump_anim_num;
        item->frame_number = anim->jump_frame_num;
```


- [x] `BaconLara_Control`:
```
        item->anim_number = g_LaraItem->anim_number;
        item->frame_number = g_LaraItem->frame_number;
```


- [x] `Bat_FixEmbeddedPosition`:
```
        item->anim_number = g_Objects[item->object_number].anim_index;
        item->frame_number = g_Anims[item->anim_number].frame_base;
        bounds = Item_GetBoundsAccurate(item);

        item->anim_number = old_anim;
        item->frame_number = old_frame;
```

- [x] `SkateKid_Draw`:
```
    item->anim_number = anim + g_Objects[O_SKATEBOARD].anim_index
        - g_Objects[O_SKATEKID].anim_index;
    Object_DrawAnimatingItem(item);
    item->anim_number = anim;
    item->frame_number = frame;
```

- [x] `Torso_Control`:
```
                g_LaraItem->anim_number = g_Objects[O_LARA_EXTRA].anim_index;
                g_LaraItem->frame_number =
                    g_Anims[g_LaraItem->anim_number].frame_base;
```

- [x] `TRex_LaraDeath`:
```
    g_LaraItem->anim_number = g_Objects[O_LARA_EXTRA].anim_index + 1;
    g_LaraItem->frame_number = g_Anims[g_LaraItem->anim_number].frame_base;
```

- [x] `Scion_Collision`:
```
        lara_item->anim_number = g_Objects[O_LARA_EXTRA].anim_index;
        lara_item->frame_number = g_Anims[lara_item->anim_number].frame_base;
```

- [x] `Scion_Collision4`:
```
        lara_item->anim_number = g_Objects[O_LARA_EXTRA].anim_index;
        lara_item->frame_number = g_Anims[lara_item->anim_number].frame_base;
```

- [x] `MidasTouch_Collision`:
```
        lara_item->anim_number = g_Objects[O_LARA_EXTRA].anim_index + 1;
        lara_item->frame_number = g_Anims[lara_item->anim_number].frame_base;
```

- [x] `MidasTouch_Collision`:
```
        lara_item->anim_number = g_Objects[O_LARA_EXTRA].anim_index;
        lara_item->frame_number = g_Anims[item->anim_number].frame_base;
```

- [x] `ThorsHandle_Control`:
```
    head_item->anim_number = g_Objects[O_THORS_HEAD].anim_index + anim;
    head_item->frame_number = g_Anims[head_item->anim_number].frame_base + frm;
```

lahm86 commented on src/game/objects/common.c:
```diff
@@ -67,8 +67,9 @@ void Object_DrawPickupItem(ITEM_INFO *item)
     // Modify item to be the anim for inv item and animation 0.
-    item->anim_number = g_Objects[item_num_option].anim_index;
-    item->frame_number = g_Anims[item->anim_number].frame_base;
+    Item_SwitchToAnim(
```
I think `Item_SwitchToAnim` will effectively set the anim number to `g_Objects[item->object_number].anim_index + g_Objects[item_num_option].anim_index` in this scenario, and the same with the frame. Is the following needed here instead?

```c
GAME_OBJECT_ID old_object_num = item->object_number;
int16_t old_frame_number = item->frame_number;
item->object_number = item_num_option;
Item_SwitchToAnim(item, 0, 0);
// ...
item->frame_number = old_frame_number;
item->object_number = old_object_num ;
```

walkawayy replied:
Oops I think I meant to just write `Item_SwitchToAnim(item, 0, 0);`. It somehow works fine though.

I don't think we need to restore the object number though? `frame_number` is already restored.

walkawayy replied:
Well changing it to `Item_SwitchToAnim(item, 0, 0);` without restoring doesn't work. 😆

![image](https://github.com/LostArtefacts/TR1X/assets/81546780/7bd8f2e5-d1e4-4216-88e8-4da167256b2b)

walkawayy replied:
This change doesn't really belong in this commit anyway, so I reverted it and added it to the to do list above.

walkawayy commented:
Repeated collision animations should be fixed now with the animation number check. It's a lengthy if condition though.

rr- reviewed (commented):
We use also `int16_t anim = item->anim_number - g_Objects[item->object_number].anim_index == …` which is inconsistent with `g_Objects[item->object_number].anim_index + … == item->anim_number`; can we introduce `Item_TestAnim` too?

walkawayy committed:
items: add Item_TestAnimEqual function

lahm86 reviewed (approved):
LGTM. Great catch on the step-back-right animation too 👍
